x-pack/plugins/enterprise_search/public/applications/enterprise_search_content/components/search_index/pipelines/ml_inference/multi_field_selector.tsx

```diff
@@ -31,13 +31,62 @@ import { MLInferenceLogic } from './ml_inference_logic';
 
 type FieldNames = Array<{ label: string }>;
 
+const TARGET_FIELD_PLACEHOLDER_TEXT_NO_FIELDS = i18n.translate(
+  'xpack.enterpriseSearch.content.indices.pipelines.addInferencePipelineModal.steps.fields.targetField.placeholder.noFields',
+  {
+    defaultMessage: 'Select a source field',
+  }
+);
+
+const TARGET_FIELD_PLACEHOLDER_TEXT_MULTIPLE_FIELDS = i18n.translate(
+  'xpack.enterpriseSearch.content.indices.pipelines.addInferencePipelineModal.steps.fields.targetField.placeholder.multipleFields',
+  {
+    defaultMessage: 'Automatically created for multi-select',
+  }
+);
+
+const TARGET_FIELD_PLACEHOLDER_TEXT_TEXT_EXPANSION_MODEL = i18n.translate(
+  'xpack.enterpriseSearch.content.indices.pipelines.addInferencePipelineModal.steps.fields.targetField.placeholder.textExpansionModel',
+  {
+    defaultMessage: 'Automatically created',
+  }
+);
+
+const TARGET_FIELD_HELP_TEXT_DEFAULT = i18n.translate(
+  'xpack.enterpriseSearch.content.indices.pipelines.addInferencePipelineModal.steps.fields.targetField.helpText',
+  {
+    defaultMessage: 'Optional. Field name where inference results should be saved.',
+  }
+);
+
+const TARGET_FIELD_HELP_TEXT_TEXT_EXPANSION_MODEL = i18n.translate(
+  'xpack.enterpriseSearch.content.indices.pipelines.addInferencePipelineModal.steps.fields.targetField.helpTextTextExpansionModel',
+  {
+    defaultMessage: 'ELSER target fields are created automatically.',
+  }
+);
+
+const getInitialTargetFieldPlaceholderText = (isTextExpansionModelSelected: boolean) =>
+  isTextExpansionModelSelected
+    ? TARGET_FIELD_PLACEHOLDER_TEXT_TEXT_EXPANSION_MODEL
+    : TARGET_FIELD_PLACEHOLDER_TEXT_NO_FIELDS;
+
+const getTargetFieldHelpText = (isTextExpansionModelSelected: boolean) =>
+  isTextExpansionModelSelected
+    ? TARGET_FIELD_HELP_TEXT_TEXT_EXPANSION_MODEL
+    : TARGET_FIELD_HELP_TEXT_DEFAULT;
+
 export const MultiFieldMapping: React.FC = () => {
   const {
     addInferencePipelineModal: { configuration, selectedSourceFields = [] },
+    isTextExpansionModelSelected,
     sourceFields,
   } = useValues(MLInferenceLogic);
   const { ingestionMethod } = useValues(IndexViewLogic);
-  const { addSelectedFieldsToMapping, selectFields } = useActions(MLInferenceLogic);
+  const { addSelectedFieldsToMapping, selectFields, setTargetField } = useActions(MLInferenceLogic);
+  const [placeholderText, setPlaceholderText] = React.useState<string>(
```

Comment:
I tried using a plain variable for `placeholderText` but I found that in that case the component doesn't update.

![Screenshot 2023-07-26 at 10 44 03](https://github.com/elastic/kibana/assets/14224983/d54eeb20-b287-4c8c-a762-25d544e8272c)

I think it's because we're changing the variable inside the event handler `onChangeSelectedFields`.

How would you structure the component to fix this (and is this something we can do in a separate PR)?

Reply:
@sphilipse I suggested @demjened use `useState` here since he needed to update the placeholder text based on user actions (selecting or removing source fields) but it didn't seemed like overkill to put into the logic.

Reply:
Yeah, I expect this is caused by React not detecting that `selectedFieldNames` is changing. Maybe we're re-assigning the same, modified array rather than pushing a new one in the logic? 

I'll approve for now because it works, but this should be addressed at some point because it's likely to cause some interesting bugs in the future.

Reply:
Ok, we can address that while refactoring the modal (we want to break it up into smaller pieces).